Web3/Classes/Core/ABI/ABIConvertible.swift

```diff
@@ -0,0 +1,347 @@
+//
+//  ABIRepresentable.swift
+//  AppAuth
+//
+//  Created by Josh Pyles on 5/22/18.
+//
+
+import Foundation
+import BigInt
+
+/// A type that is always represented as a single SolidityType
+public protocol SolidityTypeRepresentable {
+    static var solidityType: SolidityType { get }
+}
+
+/// A type that can be converted to and from Solidity ABI bytes
+public protocol ABIConvertible {
```

Comment:
Do you want to split this into `ABIRepresentable` and `ABIInitializable`? You wouldn't lose anything and make it much more flexible. It would also meet this libraries way to handle convertible types like bytes convertible or ethereum value convertible.

Reply:
That's actually a good point now that we have types that don't quite align with Swift's native types. SolidityTuple is really only representable and not initializable for example.

Reply:
I went with `ABIEncodable` and `ABIDecodable` for now as a parallel to `Codable`, but I don't mind `Representable` and `Initializable` if you prefer that

Reply:
No that's absolutely fine.